pylint/extensions/check_elif.py

```diff
@@ -38,37 +38,27 @@ def __init__(self, linter=None):
         self._init()
 
     def _init(self):
-        self._elifs = []
-        self._if_counter = 0
+        self._elifs = {}
 
     def process_tokens(self, tokens):
-        # Process tokens and look for 'if' or 'elif'
-        for _, token, _, _, _ in tokens:
-            if token == "elif":
-                self._elifs.append(True)
-            elif token == "if":
-                self._elifs.append(False)
+        """Process tokens and look for 'if' or 'elif'"""
+        self._elifs = {
+            begin: token for _, token, begin, _, _ in tokens if token in {"elif", "if"}
+        }
 
     def leave_module(self, _: nodes.Module) -> None:
         self._init()
 
-    def visit_ifexp(self, node: nodes.IfExp) -> None:
-        if isinstance(node.parent, nodes.FormattedValue):
-            return
-        self._if_counter += 1
-
-    def visit_comprehension(self, node: nodes.Comprehension) -> None:
-        self._if_counter += len(node.ifs)
-
     @check_messages("else-if-used")
     def visit_if(self, node: nodes.If) -> None:
-        if isinstance(node.parent, nodes.If):
-            orelse = node.parent.orelse
-            # current if node must directly follow an "else"
-            if orelse and orelse == [node]:
-                if not self._elifs[self._if_counter]:
-                    self.add_message("else-if-used", node=node)
-        self._if_counter += 1
+        """Current if node must directly follow an 'else'"""
+        if (
+            isinstance(node.parent, nodes.If)
+            and node.parent.orelse == [node]
+            and (node.lineno, node.col_offset) in self._elifs
+            and self._elifs[(node.lineno, node.col_offset)] == "if"
+        ):
+            self.add_message("else-if-used", node=node, confidence=HIGH)
```

Comment:
😄

Reply:
Not having an off by one error when we miscount an If increased the confidence 😄  But really the other option are:
```python
Confidence = namedtuple("Confidence", ["name", "description"])
# Warning Certainties
HIGH = Confidence("HIGH", "No false positive possible.")
INFERENCE = Confidence("INFERENCE", "Warning based on inference result.")
INFERENCE_FAILURE = Confidence(
    "INFERENCE_FAILURE", "Warning based on inference with failures."
)
UNDEFINED = Confidence("UNDEFINED", "Warning without any associated confidence level.")

CONFIDENCE_LEVELS = [HIGH, INFERENCE, INFERENCE_FAILURE, UNDEFINED]
```
There's no inference here, so it should probably by HIGH. But also, we should probably refactor this (?) It was not used before so we might as well make it make sense. I'm not confident enough to say there will never be any false positive just because there was no inference.

Reply:
Seems like we could use a `NO_INFERENCE` level. Maybe add the to the growing list of todo's for `2.13`?

Reply:
Updated https://github.com/PyCQA/pylint/issues/5317